### PR TITLE
Enabled clang-format on C++ sources

### PR DESCRIPTION
**Pull request by nandor**
Ran clang-format on sources committed to the repository and added a CI job to ensure sources are properly formatted on each commit.
Clang-format can be installed using Linux package managers and the `git clang-format` command can be used to format all files that change with a commit.

#### Related PRs / Issues

N/A

#### UI / API Impact

No impact

#### Verilog / AGFI Compatibility

No impact

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [ ] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: .github/workflows/firesim-run-tests.yml
Comment (timsnyder-siv) [inline .github/workflows/firesim-run-tests.yml:277]: I don't think we want to depend on another unrelated project's CI container to run clang-format and not provide the same version to developers to run locally via conda.

If we are going to adopt `clang-format`, we should better integrate this into our existing development environment before merging.

We should also name the step something more descriptive than `lints` unless we're thinking of cramming all of the lint checkers into this one step.  `clang-check` or `cppcheck` are more what I would consider lint checkers, this is just enforcing a formatting standard.

**Review by timsnyder-siv — CHANGES_REQUESTED** [review]
In general, I'm all for having a well-defined C++ formatting standard and I don't have any issues with `clang-format`.  If we are going to adopt it for C++ we should add it to the conda-dependencies and document this in the dev docs before merging this.

File: .github/workflows/firesim-run-tests.yml
Comment (nandor) [reply .github/workflows/firesim-run-tests.yml]: I changed this to use a default image. Will look into conda as well.

**Comment by nandor** [comment]
I have added a small note in the driver development notes about coding style and amended the conda package list to include clang-format.

File: .github/workflows/firesim-run-tests.yml
Comment (davidbiancolin) [inline .github/workflows/firesim-run-tests.yml:302]: So i'm a little confused here -- this workflow should only be run on open, non-draft PRs against main (and 1.[0-9].x), why _isn't_ the `base_ref` always set?

Comment (nandor) [reply .github/workflows/firesim-run-tests.yml]: I've removed this check. The rule was adapted from another CI job which did not run only on pull events that expose base_ref. However, if you think the triggers of this job will change, we should revert.

File: sim/midas/src/main/cc/bridges/synthesized_prints.h
Comment (timsnyder-siv) [inline sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: I'll be the first to admit that the signature of this constructor is comically long but `clang-format` has made this almost unreadable by smashing collapsing it.

**Review by timsnyder-siv — COMMENTED** [review]
I've taken a more serious look at the actual output of `clang-format` and have some thoughts for discussion.  In general, I'm really happy to see `clang-format` being used because it is going to make the code more consistent. Having a clearly defined format that is expected in the repo makes it possible to reformat it to anyone's individual liking while editing (and then reapply the official formatting).  Can be done either with git's smudge filters or with editor support (each has it's plusses and minuses).

**Comment by fabianschuiki** [comment]
This is very nice! I love the idea of establishing _any_ `clang-format` style guidelines for the C++ here. I don't have any strong opinions on the individual style options. My suggestion would be to settle on _any_ initial formatting setup, and then iterate on the formatting details afterwards with subsequent PRs -- after all it just takes a run of clang-format across all sources in this repo to update to a new style.

File: sim/firesim-lib/src/main/cc/bridges/blockdev.h
Comment (nandor) [reply sim/firesim-lib/src/main/cc/bridges/blockdev.h:20]: I strongly disagree with per-project policies instead of language-specific policies. For C++, 2 spaces and 80 characters are quite common and enforced in other related projects. I find Scala very hard to read since it exceeds 100 characters per line (which I think the limit might be in FireSim). Keeping indentation to 2 spaces and line length to 80 characters results, in my opinion, in a better overview of the code.

File: sim/midas/src/main/cc/bridges/synthesized_prints.h
Comment (nandor) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: I think we can adjust this in clang-format.

File: sim/firesim-lib/src/main/cc/bridges/dromajo.h
Comment (nandor) [reply sim/firesim-lib/src/main/cc/bridges/dromajo.h:32]: I'll keep this in mind if I have to change code involving trailing underscores in the future.

File: scripts/machine-launch-script.sh
Comment (fabianschuiki) [reply scripts/machine-launch-script.sh:285]: :+1: That sounds great!

File: sim/firesim-lib/src/main/cc/bridges/dromajo.cc
Comment (fabianschuiki) [reply sim/firesim-lib/src/main/cc/bridges/dromajo.cc:278]: Totally agree. I've heard Dennis Ritchie being quoted as saying that the worst mistake he made in his entire life was making braces optional in C :smile:

File: sim/firesim-lib/src/main/cc/bridges/dromajo.cc
Comment (nandor) [reply sim/firesim-lib/src/main/cc/bridges/dromajo.cc:33]: I'd discourage multiple pointer declarations on a single line. In general, if a pointer declaration is not trivial, typedefs with comments should be inserted to clarify their meaning and then the location of the pointer or the reference is irrelevant. 

Regarding the Chromium style, I would very much prefer sticking to the LLVM style.

File: sim/firesim-lib/src/main/cc/bridges/dromajo.h
Comment (fabianschuiki) [reply sim/firesim-lib/src/main/cc/bridges/dromajo.h:32]: Should we spin this off into a dedicated tracking issue for making names more consistent? I like your suggestion of keeping this PR minimal and to the point of just adding consistent automated formatting.

File: sim/firesim-lib/src/main/cc/bridges/blockdev.h
Comment (fabianschuiki) [reply sim/firesim-lib/src/main/cc/bridges/blockdev.h:20]: My suggestion here would be to not necessarily condition one on the other. C++ can have a significant rightward drift if lambdas are used excessively, which would suggest 2 spaces of indentation. But I could also see other languages having 4 spaces as the idiomatic indentation, which is totally fine. And you might have languages like Make that rely on tabs, so unifying this across language boundaries seems like it's more trouble than benefit. We should definitely stay consistent within languages though.

File: sim/midas/src/main/cc/bridges/synthesized_prints.h
Comment (fabianschuiki) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: We could add
```
BinPackArguments: false
BinPackParameters: false
```
to ensure func args are either all on one line, or all on their own line.

File: sim/firesim-lib/src/main/cc/bridges/dromajo.cc
Comment (fabianschuiki) [reply sim/firesim-lib/src/main/cc/bridges/dromajo.cc:33]: Totally agree with @timsnyder-siv here: we should probably just go with the `PointerAlignment: right` that the LLVM style uses by default. No need to switch to Chromium style; there's two simple bin packing options we can add to get the func args right.

File: sim/midas/src/main/cc/bridges/synthesized_prints.h
Comment (nandor) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: I tried these options and the output doesn't look great. We can either roll with what we have now or I can try to find some other options, although that would introduce further delays since I don't quite have time for that now.

File: sim/firesim-lib/src/main/cc/bridges/blockdev.h
Comment (timsnyder-siv) [reply sim/firesim-lib/src/main/cc/bridges/blockdev.h:20]: @davidbiancolin confirmed that the scala is using indent of 2 spaces.  so the indenting comment I made here is moot.  I was primarily concerned about Scala and C++ because it is most likely that one would be looking at both those at the same time in this project.  I should have been more precise and limited in my comment.

I'm fine with shifting it to 2 spaces.  I used to like 2 spaces more 20 years ago. As my eyes have gotten older, I like it less. I should probably suck it up and increase the font size on my display again or go to the eye doctor. 🤣 

Excessive use of lambdas should probably be avoided, as should signatures that take a page of code to represent without binpacking the arguments (or 10 lines with them packed...).

David also mentioned that the Scala column limit is currently 120.  That seems excessive to me but the coming adoption of scalafmt will make it easier to reflow that on the fly.

File: sim/midas/src/main/cc/bridges/synthesized_prints.h
Comment (timsnyder-siv) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: @nandor can you be more specific about what didn't look great?  When I scanned through the output with those options enabled, it looked like the output was only changing the arguments to be either all on one line or all on their own line.

**Comment by davidbiancolin** [comment]
For reference, the scalaFmt is here: https://github.com/firesim/firesim/blob/main/sim/.scalafmt.conf#L210

**Comment by davidbiancolin** [comment]
Really the only two things i'm willing to bike shed about are [the classic] column limit (maybe 100 instead of 80?), and parameter list formatting (see above). 

It's worth noting though, that most of the instances of long parameter + initializer lists are in bridge definitions will disappear once we make some of the driver modifications we've proposed, since these parameters will be captured in automatically constructed instances which will be passed to constructors their stead.

**Comment by davidbiancolin** [comment]
Finally, I am a little reluctant to fiddle with the formatting if the pertubatios are going to introduce +/- 1000 line PRs that i'll have to deal with in our fork. I'd prefer to just pick something and run with it.

File: sim/midas/src/main/cc/bridges/synthesized_prints.h
Comment (nandor) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]:
```
fprintf(stderr,
          "Some format %x\n",
          number);
```
is one example

Comment (nandor) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: We could bin-pack arguments to function calls, placing parameters on new lines.

Comment (timsnyder-siv) [reply sim/midas/src/main/cc/bridges/synthesized_prints.h:53]: Thanks for providing the detail.

**Comment by nandor** [comment]
Do you have a decision on parameter formatting and column widths?

File: scripts/machine-launch-script.sh
Comment (timsnyder-siv) [inline scripts/machine-launch-script.sh:285]: Did some more looking and found it.  They put it in the `clang-tools` package.  I'll ask about splitting it out into the clang-format package so that we don't have to install the rest of clang but in the meantime, we could:
```suggestion
    CONDA_PACKAGE_SPECS+=( clang-format clang-tools )
```

**Comment by davidbiancolin** [comment]
My final call is to add:
```
BinPackArguments: false
BinPackParameters: false
```
We leave the column limit at 80. We can circle back on the BinPackingX when we've cleaned up the header situation.

File: docs/Developer-Docs/GoldenGate-and-Driver-Development.rst
Comment (timsnyder-siv) [inline docs/Developer-Docs/GoldenGate-and-Driver-Development.rst:163]:
```suggestion
The C++ sources are formatted using ``clang-format`` and all submitted pull-requests 
must be formatted following `https://github.com/firesim/firesim/blob/main/.clang_format`_ 
prior to being accepted and merged.

``git clang-format`` can be used before committing to ensure that files are properly formatted. 
```

RST inline literals need to be surrounded in double backtics.

Should explain that we're using `clang-format` in a way that doesn't imply that we've adopted something called the "LLVM style guide" that lives somewhere like https://llvm.org/docs/CodingStandards.html or something else that a search engine might confuse someone with.

**Review by timsnyder-siv — APPROVED** [review]

**Comment by davidbiancolin** [comment]
Im running out of time to play with this tonight, but i'll get to it tomorrow.

**Comment by davidbiancolin** [comment]
I've decided that the argument packing to `fprintf` isn't triggering enough for me to go and add waivers to metasimulation harnesses. We should stop using `fprintf` in favor of C++ I/O anyways.

**Pull request merged**
